### PR TITLE
Compressed histogram V2 support

### PR DESCRIPTION
**Pull request by filipecosta90**
This is a WIP PR to share the current status of compressed histogram v2 encoding and decoding. wanted to pick your brain @ahothan on where I might be failing on the porting. 

## current to be added APIs
- `func Decode(encoded []byte) (rh *Histogram, err error)` -  Decode returns a new Histogram by decoding it from a String containing a base64 encoded compressed histogram representation. 
- `func (h *Histogram) Encode(version int32) (buffer []byte, err error)` -  Encode returns a snapshot view of the Histogram. The snapshot is compact binary representations of the state of the histogram. They are intended to be used for archival or transmission to other systems for further analysis. 

## quick notes 
Please notice that:
- this is still in draft mode, still to be linted and properly documented.
- By using the [decoding-demo](https://hdrhistogram.github.io/HdrHistogramJSDemo/decoding-demo.html) we're able to quickly check that the go version encoded histograms are being validly parsed by the js version.

## current issue
The current issue to be discussed is that taking as examples the encoded histograms from JS, python,etc... we're able to decode them, but if we encode them again the output based64 is different from the original one. Here is a good example of it:
- Original base64 input: HISTFAAAAB542pNpmSzMwMDAxAABzFCaEUoz2X+AMIKZAEARAtM=
- Output base64 encoded: HISTFAAAACZ42pJpmSzMwMDAxMDAwMjAwMDMAAGMUJoJxg9mAgQAAP//NU0Bpg==
        
```
GO111MODULE=on go test -count=1 ./...
--- FAIL: TestHistogram_Load (0.00s)
    hdr_encoding_test.go:27: The input and decoded->encoded representations differ:
        Original base64 input: HISTFAAAAB542pNpmSzMwMDAxAABzFCaEUoz2X+AMIKZAEARAtM=
        Output base64 encoded: HISTFAAAACZ42pJpmSzMwMDAxMDAwMjAwMDMAAGMUJoJxg9mAgQAAP//NU0Bpg==
        Differences (-got +want):
          bytes.Join({
                "HISTFAAAA",
        -       "B542pNpmSzMwMDAxAABzFCaEUoz2X+AMIKZAEARAtM",
        +       "CZ42pJpmSzMwMDAxMDAwMjAwMDMAAGMUJoJxg9mAgQAAP//NU0Bpg=",
                "=",
          }, "")
FAIL
FAIL    github.com/HdrHistogram/hdrhistogram-go 0.229s
FAIL
make: *** [test] Error 1
```

### CODE REVIEW

**Comment by ahothan** [comment]
Thanks for the PR, I'll try to spend time on this sometime this week,
I will try to port some of the test cases I have in python version that create a histogram from scratch and verify that the encoded histogram are similar. Once the encoding is verified we can have a closer look at the decoding side,
Clearly decode+encode should be idempotent.

**Comment by ahothan** [comment]
We should probably only support v2 format, don't think anybody is using V1 any more.

File: hdr.go [hunk 1]
```diff
@@ -455,6 +463,10 @@ func (h *Histogram) countsIndexFor(v int64) int {
 	return int(h.countsIndex(bucketIdx, subBucketIdx))
 }
 
+func (h *Histogram) getIntegerToDoubleValueConversionRatio() int64 {
```
Comment (ahothan) [inline hdr.go:466]: this field must be encoded as float64, 
see https://github.com/HdrHistogram/HdrHistogram/blob/e9e3d3ea7bfbc628863c7d20fd96489813690d84/src/main/java/org/HdrHistogram/AbstractHistogram.java#L1998

Comment (ahothan) [reply hdr.go:466]: with this change I can at least decode the resulting encoded string of an empty histogram:
`func TestEmpty(t *testing.T) {
	var lowest int64 = 1
	var highest int64 = 3600 * 1000 * 1000
	sigfigs := 3
	h := hdrhistogram.New(lowest, highest, sigfigs)
	bufferOut, _ := h.Encode(hdrhistogram.V2CompressedEncodingCookieBase)
	fmt.Printf("Empty histoblob: %s\n", string(bufferOut))
}`
yields:
Empty histoblob: HISTFAAAACR42pJpmSzMwMDACMXMDBAAYjNcm7yEwf4DVAQQAAD//1WhBIk=

which I can decode with python and I can verify it is equal to the same empty histogram.

Comment (filipecosta90) [reply hdr.go]: Thank you for finding this issue @ahothan. Fixed in the latest commit.

**Comment by ahothan** [comment]
Looks good.
The reason the recode of a decoded histoblob coming from another implementation is different is due to the difference in compression library implementation. As long as they can decode properly, it should be fine.

**Pull request merged**